contracts/FeeProvider/FeeCollector.sol

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./FeeDealProvider.sol";
+import "./FeeLockProvider.sol";
+import "./FeeTimedProvider.sol";
+
+contract FeeCollector is IERC721Receiver, IFeeCollector {
+    uint256 public fee; // 1e18 = 100%
+    address public immutable feeCollector;
+    ILockDealNFT public immutable lockDealNFT;
+    IProvider public immutable feeDealProvider;
+    IProvider public immutable feeLockProvider;
+    IProvider public immutable feeTimedProvider;
+    bool public feeCollected;
+
+    constructor(uint256 _fee, address _feeCollector, ILockDealNFT _lockDealNFT) {
+        fee = _fee;
+        feeCollector = _feeCollector;
+        lockDealNFT = _lockDealNFT;
+        feeDealProvider = new FeeDealProvider(IFeeCollector(this), _lockDealNFT);
+        feeLockProvider = new FeeLockProvider(_lockDealNFT, feeDealProvider);
+        feeTimedProvider = new FeeTimedProvider(_lockDealNFT, feeLockProvider);
```

Comment:
Lets make each to take the _lockDealNFT from the Deal - not pass here, only in first